File: types/merge_intervals/Merge_Intervals.cpp
using namespace std;

#include <algorithm>
#include <iostream>
#include <vector>

class Interval {
 public:
  int start = 0;
  int end = 0;

  Interval(int start, int end) {
    this->start = start;
    this->end = end;
  }
};

class MergeIntervals {
 public:
  static vector<Interval> merge(vector<Interval> &intervals) {
    vector<Interval> mergedIntervals;
    Interval current_interval(0,0);
    int start, end;


    // Sort the list of intervals based on the start times
    sort(intervals.begin(), 
            intervals.end(), 
              [](Interval &x, Interval &y) { return x.start < y.start; } );



    vector<Interval>::iterator itr = intervals.begin();
    start = (*itr).start;
    end = (*itr).end;
    itr++; 
    while(itr != intervals.end()) {

      current_interval = *itr;

      if(current_interval.start < end) { // Overlapping intevals

        end = max(current_interval.end, end);

      } else { // Broken intervals

        mergedIntervals.push_back({start,end}); // Push the connected interval

        start = current_interval.start; // Start from the next interval
        end = current_interval.end;

      }

      itr++;

    }


    mergedIntervals.push_back({start,end});

    return mergedIntervals;
  }
};

int main(int argc, char *argv[]) {
  vector<Interval> input = {{1, 3}, {2, 5}, {7, 9}};
  cout << "Merged intervals: ";
  for (auto interval : MergeIntervals::merge(input)) {
    cout << "[" << interval.start << "," << interval.end << "] ";
  }
  cout << endl;

  input = {{6, 7}, {2, 4}, {5, 9}};
  cout << "Merged intervals: ";
  for (auto interval : MergeIntervals::merge(input)) {
    cout << "[" << interval.start << "," << interval.end << "] ";
  }
  cout << endl;

  input = {{1, 4}, {2, 6}, {3, 5}};
  cout << "Merged intervals: ";
  for (auto interval : MergeIntervals::merge(input)) {
    cout << "[" << interval.start << "," << interval.end << "] ";
  }
  cout << endl;
}

File: types/merge_intervals/Merge_Intervals.py

#             ____________   ___________  _________________________________  ________________   ______   _____    ______
intervals =  [[1,3], [2,4], [5,7], [6,8], [9,13], [13,15], [14,15], [14,19], [25,26], [26,27], [40,47], [50,60], [70,80]]   # Sorted based on the start time

output = []


start = intervals[0][0]
end = intervals[0][1]

for i in range(len(intervals)):

    if i == 0:
        continue

    if intervals[i][0] <= end:   # Current start < end # Intervals are overlapping

        end = max(end, intervals[i][1]) # Merge them by updating the end time

    else:  # Intervals are broken

        output.append( [start, end] )

        start = intervals[i][0]
        end = intervals[i][1]

output.append([start, end])

print(output) # [[1, 4], [5, 8], [9, 19]]

